src/Validation.elm

```diff
@@ -36,49 +50,81 @@ import Zipper
 --     ( lp ++ error [] (f z), z )
 
 
-isCorrectTableau : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
-isCorrectTableau z =
+isCorrectTableau : Config -> Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+isCorrectTableau config z =
     Errors.merge2 (always2 z)
-        (isCorrectNode z)
+        (isCorrectNode config z)
         (List.foldl
             (Errors.merge2 (always2 z))
             (Ok z)
-            (List.map isCorrectTableau (Zipper.children z))
+            (List.map (isCorrectTableau config) (Zipper.children z))
         )
 
 
 isValidNode : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
 isValidNode z =
-    Errors.merge3 (always3 z)
+    Errors.merge4 (always4 z)
         (isValidFormula z)
         (isValidNodeRef z)
+        (isValidSubstitution z)
         (areValidCloseRefs z)
 
 
-isCorrectNode : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
-isCorrectNode z =
+isCorrectNode : Config -> Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+isCorrectNode config z =
     isValidNode z
         |> Result.andThen
             (\_ ->
                 Errors.merge2 second
-                    (isCorrectRule z)
+                    (isCorrectRule config z)
                     (areCorrectCloseRefs z)
             )
 
 
 {-| Just for the formula dislplay -- don't check the ref for syntax
 -}
-isCorrectFormula : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
-isCorrectFormula z =
+isCorrectFormula : Config -> Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+isCorrectFormula config z =
     isValidFormula z
-        |> Result.andThen isCorrectRule
+        |> Result.andThen (isCorrectRule config)
 
 
 isValidFormula : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
 isValidFormula z =
     z |> Zipper.zNode |> .formula |> Result.mapError (parseProblem z) |> Result.map (always z)
 
 
+areTermsDistinct : Term.Substitution -> Bool
+areTermsDistinct subst =
+    let
+        terms =
+            Dict.values subst
+    in
+    List.length terms == Set.size (Set.fromList (List.map Term.toString terms))
+
+
+isValidSubstitution : Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+isValidSubstitution z =
+    if Zipper.up z == z then
+        Ok z
+
+    else
+        case Zipper.zSubstitution (Zipper.up z) of
+            Just subst ->
+                subst
+                    |> .parsedSubst
+                    |> Result.mapError (\_ -> syntaxProblem z "Wrong form of substitution")
+                    |> Result.map (\parsedS -> Dict.union parsedS (implicitSubst parsedS z))
+                    |> Result.andThen
+                        (checkPredicate areTermsDistinct
+                            (syntaxProblem z "Substituted terms must be distict")
```

Comment:
Substituted terms need not be distinct in the γ* rule.
Substituted _variables_ must be distinct in the δ* rule.
Also: <code>disti*n*ct</code>.